src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -0,0 +1,262 @@
+package org.folio.auth.authtokenmodule.tokens;
+
+import io.vertx.core.Future;
+import io.vertx.core.MultiMap;
+import io.vertx.core.http.HttpServerRequest;
+import io.vertx.core.json.JsonObject;
+import java.util.Base64;
+import com.nimbusds.jose.JOSEException;
+import org.apache.commons.lang3.StringUtils;
+import org.folio.auth.authtokenmodule.BadSignatureException;
+import org.folio.auth.authtokenmodule.TokenCreator;
+import org.folio.okapi.common.XOkapiHeaders;
+import java.text.ParseException;
+
+/**
+ * An abstract class that all token types should extend, with one required method
+ * that all must implement to handle their validation logic. This class also exposes
+ * a handful of static methods which provide convenient ways of working with tokens,
+ * including a method to parse and validate a token in one step.
+ */
+public abstract class Token {
+  private static final String CALLING_MODULE_HEADER = "X-Okapi-Calling-Module";
+  private boolean usesDummyPermissionsSource;
+  protected static final String UNDEFINED_USER_NAME = "UNDEFINED_USER__";
```

Comment:
This also bugs me as a hack. Not your fault though. I wished it could be removed.

Reply:
@adamdickmeiss you mean the whole `UNDEFINED _USER__` thing? My sense is that this is needed to handle the pre-auth state (before login). But I haven't dug into it much. Just tried to keep my code from breaking the existing functionality. But I agree it doesn't smell right.